2.0/m_bannegate.cpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * InspIRCd -- Internet Relay Chat Daemon
+ *
+ *   Copyright (C) 2017 Dylan Frank <b00mx0r@aureus.pw>
+ *
+ * This file is part of InspIRCd.  InspIRCd is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+/* $ModAuthor: Dylan Frank */
+/* $ModAuthorMail: b00mx0r@aureus.pw */
+/* $ModDesc: Enables ban negation by prefixing the match sequence with ~ */
+/* $ModDepends: core 2.0 */
+
+#include "inspircd.h"
+
+class ModuleBanNegate : public Module
+{
+ public:
+ 	void init()
+ 	{
+ 		ServerInstance->Modules->Attach(I_OnCheckBan, this);
+ 	}
+
+ 	void Prioritize()
+ 	{
+ 		ServerInstance->Modules->SetPriority(this, I_OnCheckBan, PRIORITY_FIRST);
+ 	}
+
+	ModResult OnCheckBan(User *source, Channel *chan, const std::string& mask)
+	{
+		// If our matching mask begins with the negate charater, but does not have multiple in a row (to avoid nested loops)
+		if (mask[0] == '~' && mask[1] != '~')
```

Comment:
You should check the length here.

Reply:
This isn't meant to only be for extbans. You can also do +b ~nick!*@* to ban everyone except nick!*@*. While that probably won't be used much, this flexibility is why I love inspircd modules.

Reply:
Yes but theres no guarantee that `mask[1]` will actually be set and if its not this will cause a crash.

Reply:
Right, thanks for catching that! Good to know.